src/net/telemetry.rs

```diff
@@ -0,0 +1,168 @@
+pub struct TelemetryBuffer {
+    /// The latest input sample on ADC0/ADC1.
+    pub latest_samples: [i16; 2],
+    /// The latest output code on DAC0/DAC1.
+    pub latest_outputs: [u16; 2],
+    /// The latest digital input states during processing.
+    pub digital_inputs: [bool; 2],
+}
+
+/// The telemetry structure is data that is ultimately reported as telemetry over MQTT.
+///
+/// # Note
+/// This structure should be generated on-demand by the buffer when required to minimize conversion
+/// overhead.
+#[derive(Serialize)]
+pub struct Telemetry {
+    input_levels: [f32; 2],
+    output_levels: [f32; 2],
+    digital_inputs: [bool; 2],
+}
+
+impl Default for TelemetryBuffer {
+    fn default() -> Self {
+        Self {
+            latest_samples: [0, 0],
+            latest_outputs: [0, 0],
+            digital_inputs: [false, false],
+        }
+    }
+}
+
+impl TelemetryBuffer {
+    /// Convert the telemetry buffer to finalized, SI-unit telemetry for reporting.
+    ///
+    /// # Args
+    /// * `afe0` - The current AFE configuration for channel 0.
+    /// * `afe1` - The current AFE configuration for channel 1.
+    ///
+    /// # Returns
+    /// The finalized telemetry structure that can be serialized and reported.
+    pub fn finalize(self, afe0: AfeGain, afe1: AfeGain) -> Telemetry {
+        // The input voltage is measured by the ADC across a dynamic scale of +/- 4.096 V with a
+        // dynamic range across signed integers. Additionally, the signal is fully differential, so
+        // the differential voltage is measured at the ADC. Thus, the single-ended signal is
+        // measured at the input is half of the ADC-reported measurement. As a pre-filter, the
+        // input signal has a fixed gain of 1/5 through a static input active filter. Finally, at
+        // the very front-end of the signal, there's an analog input multiplier that is
+        // configurable by the user.
+        let in0_volts =
+            (self.latest_samples[0] as f32 / i16::MAX as f32) * 4.096 / 2.0
+                * 5.0
+                / afe0.as_multiplier();
+        let in1_volts =
+            (self.latest_samples[1] as f32 / i16::MAX as f32) * 4.096 / 2.0
+                * 5.0
+                / afe1.as_multiplier();
+
+        // The output voltage is generated by the DAC with an output range of +/- 4.096 V. This
+        // signal then passes through a 2.5x gain stage. Note that the DAC operates using unsigned
+        // integers, and u16::MAX / 2 is considered zero voltage output. Thus, the dynamic range of
+        // the output stage is +/- 10.24 V. At a DAC code of zero, there is an output of -10.24 V,
+        // and at a max DAC code, there is an output of 10.24 V.
+        let out0_volts = (10.24 * 2.0)
+            * (self.latest_outputs[0] as f32 / (u16::MAX as f32))
+            - 10.24;
+        let out1_volts = (10.24 * 2.0)
+            * (self.latest_outputs[1] as f32 / (u16::MAX as f32))
+            - 10.24;
+
```

Comment:
Let's collect the scaling factors.

```suggestion
        let volt_per_lsb = (5.0 * 4.096 / 2.0 / i16::MAX as f32);
        let in0_volts =
            self.latest_samples[0] as f32 * (volt_per_lsb / afe0.as_multiplier());
        let in1_volts =
            self.latest_samples[1] as f32 * (volt_per_lsb / afe1.as_multiplier());

        // The output voltage is generated by the DAC with an output range of +/- 4.096 V. This
        // signal then passes through a 2.5x gain stage. Note that the DAC operates using unsigned
        // integers, and u16::MAX / 2 is considered zero voltage output. Thus, the dynamic range of
        // the output stage is +/- 10.24 V. At a DAC code of zero, there is an output of -10.24 V,
        // and at a max DAC code, there is an output of 10.24 V.
        let volt_per_lsb = (10.24 * 2.0 / u16::MAX as f32);
        let out0_volts = (self.latest_outputs[0] ^ 0x8000) as i16 as f32 * volt_per_lsb;
        let out1_volts = (self.latest_outputs[1] ^ 0x8000) as i16 as f32 * volt_per_lsb;
```

Reply:
Just a heads up that the DAC calculation is actually a linear transform with an offset of -10.24 volts, so there's both a slope + offset component, whereas the ADCs have just a slope component.

Reply:
But the XOR of the MSB is right there in the table. That's not a lot of math.
If the XOR is frightening, and to be closer to your understanding you can also do `code.wrapping_add(0x8000)` or `code.wrapping_sub(0x8000)`. All three are exactly the same.

Reply:
Calculation based on decimal constants like `32767` are really tricky and so frequently riddled with problems/bugs. Remember all the code-to-SI off-by-one conversions in Booster, Pounder or https://github.com/quartiq/stabilizer/pull/341#discussion_r628054750. These are binary systems. Why talk to them in decimal numbers?

Reply:
And to make my point: the same error here again. The conversion isn't correct. The DAC scale isn't `10.24 * 2.0 / u16::MAX as f32` but `4.096 * 2.5 / (1 << 15) as f32`. That's different.

Reply:
And finally also the ADC scale isn't `5.0 / 2.0 * 4.096 / i16::MAX as f32` but `5.0 * 4.096 / 2.0 / (1 << 15) as f32`. Again off-by-one.

Reply:
Just for clarities sake, you're neglecting to incorporate the offset voltage, which corresponds to the MSB. In any case, I actually think the XOR is more appropriate as I failed to note originally that the MAX value is actually a step below full-scale, so I'm updating to use that as well (especially since this will all be encapsulated now)